### PR TITLE
do not scale gradient in bf16 mode

### PR DESCRIPTION
**Pull request by kashif**
# What does this PR do?

Turn off gradient scaling in the trainer when bf16 mode is selected. Only use gradient scaling in float16 mode.

## Who can review?

 @sgugger and @stas00

### CODE REVIEW

**Review by stas00 — APPROVED** [review]
Thank you, Kashif. This has been long overdue!

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -595,7 +595,8 @@ def __init__(
             if args.half_precision_backend == "cuda_amp":
                 self.use_cuda_amp = True
                 self.amp_dtype = torch.float16 if args.fp16 else torch.bfloat16
-                self.do_grad_scaling = True
+                #  bf16 does not need grad scaling
+                self.do_grad_scaling = self.amp_dtype == torch.float16
```
Comment (sgugger) [inline src/transformers/trainer.py:599]: Yes to this, and I think we can put the big test below under `if self.do_grad_scaling`.

**Review by sgugger — APPROVED** [review]
Thanks for working on this! I think we can clean up a tiny bit more the code but this is the crux of the issue.

File: src/transformers/trainer.py (hunk 1 above)
Comment (kashif) [reply src/transformers/trainer.py:599]: thanks! added

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -595,27 +595,29 @@ def __init__(
...
-                        from torch.distributed.fsdp.sharded_grad_scaler import (
-                            ShardedGradScaler as FSDPShardedGradScaler,
-                        )
+                #  bf16 does not need grad scaling
+                self.do_grad_scaling = self.amp_dtype == torch.float16
+                if self.do_grad_scaling:
+                    if self.sharded_ddp is not None:
+                        self.scaler = ShardedGradScaler()
+                    elif self.fsdp is not None:
+                        if self.amp_dtype == torch.float16:
```
Comment (sgugger) [inline src/transformers/trainer.py:604]:
```suggestion
                        if self.amp_dtype == torch.float16:
```
Can now be removed.

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -595,27 +595,29 @@ def __init__(
...
-                        self.scaler = FSDPShardedGradScaler()
-                    else:
-                        self.do_grad_scaling = False
-                        self.use_cuda_amp = False
-                        self.amp_dtype = None
+                            self.scaler = FSDPShardedGradScaler()
+                        else:
+                            self.do_grad_scaling = False
+                            self.use_cuda_amp = False
+                            self.amp_dtype = None
```
Comment (sgugger) [inline src/transformers/trainer.py:613]: Just realized there is this else block here. Clearly `self.do_grad_scaling = False` is not necessary, but you might need to have the two other lines somewhere else.

@pacman100 FSDP doesn't handle bfloat16 at all?

**Review by sgugger — APPROVED** [review]
Perfect, thanks!

File: src/transformers/trainer.py (hunk 3 above)
Comment (pacman100) [reply src/transformers/trainer.py]: Hello @sgugger, similar to DeepSpeed, FSDP also manages their own half-precision, however, for FP16 it needs `ShardedGradScaler`.  Here's an example notebook from PyTorch team wrt FSDP MixedPrecision: https://github.com/lessw2020/transformer_central/blob/main/mixed_precision/mixed_precision_fsdp.ipynb

**Pull request merged**